File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.py
def solution(new_id):
    new_id = new_id.lower()
    new_id = list(new_id)

    for i in range(len(new_id)):
        if new_id[i].isalnum() or new_id[i] in "-_.":
            continue
        else:
            new_id[i] = ''

    new_id = ''.join(new_id)

    while ".." in new_id:
        new_id = new_id.replace("..", ".")

    if new_id and new_id[0] == ".":
        new_id = new_id[1:]

    if new_id and new_id[-1] == ".":
        new_id = new_id[:-1]

    if new_id == "":
        new_id = "a"

    if len(new_id) >= 16:
        new_id = new_id[:15]
        if new_id[-1] == ".":
            new_id = new_id[:-1]

    if len(new_id) <= 2:
        while len(new_id) < 3:
            new_id += new_id[-1]

    return new_id

File: 프로그래머스/1/72410. 신규 아이디 추천/신규 아이디 추천.java
class Solution {
    public String solution(String new_id) {
        // 1
        new_id = new_id.toLowerCase();

        // 2
        StringBuilder sb = new StringBuilder();
        for (Character c : new_id.toCharArray()) {
            if (Character.isDigit(c) || Character.isAlphabetic(c) || c == '-' || c == '_' || c == '.') {
                sb.append(c);
            }
        }
        new_id = sb.toString();
        System.out.println(new_id);

        // 3
        while (new_id.contains("..")) {
            new_id = new_id.replace("..", ".");
        }
        System.out.println(new_id);

        // 4
        if (new_id.startsWith(".")) {
            new_id = new_id.substring(1);
        }
        if (new_id.endsWith(".")) {
            new_id = new_id.substring(0, new_id.length() - 1);
        }
        System.out.println(new_id);

        // 5
        if (new_id.isEmpty()) {
            new_id = "a";
        }
        System.out.println(new_id);

        // 6
        if (new_id.length() >= 16) {
            new_id = new_id.substring(0, 15);
            if (new_id.endsWith(".")) {
                new_id = new_id.substring(0, new_id.length() - 1);
            }
        }
        System.out.println(new_id);

        // 7
        if (new_id.length() <= 2) {
            while(new_id.length() < 3) {
                new_id += new_id.charAt(new_id.length() -1);
            }
        }
        System.out.println(new_id);

        return new_id;
    }
}